Userland/Libraries/LibGL/Clipper.cpp

```diff
@@ -0,0 +1,111 @@
+    //  n · [P1 + ((P2 - P1) * t) - plane] = 0.
+    //  Substitude seg_length = P2 - P1 (length of line segment) and dist = P1 - plane (distance from P1 to plane)
+    //
+    //  By rearranging, we now end up with
+    //
+    //  n·[P1 + (seg_length * t) - plane] = 0
+    //  n·(dist) + n·(seg_length * t) = 0
+    //  n·(seg_length*t) = -n·(dist)
+    //  Therefore
+    //  t = (-n·(dist))/(n·seg_length)
+    //
+    // NOTE: n is the normal vector to the plane we are clipping against.
+    //
+    // Proof of algorithm found here
+    // http://studymaterial.unipune.ac.in:8080/jspui/bitstream/123456789/4843/1/Cyrus_Beck_Algo.pdf
+    // And here (slightly more intuitive with a better diagram)
+    // https://www.csee.umbc.edu/~rheingan/435/pages/res/gen-5.Clipping-single-page-0.html
+
+
+    FloatVector4 seg_length = vec - prev_vec;   // P2 - P1
+    FloatVector4 dist = prev_vec - clip_planes[plane_index]; // Distance from "out of bounds" vector to plane
+
+    float plane_normal_line_segment_dot_product = clip_plane_normals[plane_index].dot(seg_length);
+    float neg_plane_normal_dist_dot_procut = -clip_plane_normals[plane_index].dot(dist);
+    float t = (plane_normal_line_segment_dot_product / neg_plane_normal_dist_dot_procut);
+
+    // P(t) = P1 + (t * (P2 - P1))
+    FloatVector4 lerped_vec = prev_vec + (seg_length * t);
+
+    return lerped_vec;
+}
+
+// FIXME: Getting too close to zNear causes VERY serious artifacting. Should we cull the whole triangle??
+void GL::Clipper::clip_triangle_against_frustum(Vector<FloatVector4>& input_verts)
+{
+    Vector<FloatVector4, 6> clipped_polygon;
+    Vector<FloatVector4, 6> input = input_verts;
+    Vector<FloatVector4, 6>* current = &clipped_polygon;
+    Vector<FloatVector4, 6>* output_list = &input;
+    ScopeGuard guard { [&] { input_verts = *output_list; } };
+
+    for (u8 plane = 0; plane < NUMBER_OF_CLIPPING_PLANES; plane++) {
+        swap(current, output_list);
+        clipped_polygon.clear();
+
+        if (current->size() == 0) {
+            return;
+        }
+
+        // Save me, C++23
+        for (size_t i = 0; i < current->size(); i++) {
+            const auto& curr_vec = current->at(i);
+            const auto& prev_vec = current->at((i - 1) % current->size());
+
+            if (point_within_clip_plane(curr_vec, static_cast<ClipPlane>(plane))) {
+                if (!point_within_clip_plane(prev_vec, static_cast<ClipPlane>(plane))) {
+                    FloatVector4 intersect = clip_intersection_point(prev_vec, curr_vec, static_cast<ClipPlane>(plane));
+                    clipped_polygon.append(intersect);
+                }
+
+                clipped_polygon.append(curr_vec);
+            } else if (point_within_clip_plane(prev_vec, static_cast<ClipPlane>(plane))) {
+                FloatVector4 intersect = clip_intersection_point(prev_vec, curr_vec, static_cast<ClipPlane>(plane));
+                clipped_polygon.append(intersect);
+            }
+        }
+    }
+}
```

Comment:
These changes seem to have reduced CPU utilisation of the library by around 30%!!!!!

Reply:
What exactly do you mean by this, are you stating that the CPU time spent on each frame has decreased by 30%, or that the CPU utilization percentage has decreased by 30%?

Reply:
CPU usage as measured in `System Monitor` has reduced on average 30%, from 80 to about 50.

With regards to actually profiling, the amount of calls to `malloc`/`free` inside the clipping functions has gone from ~40% of execution time to 16%.

Reply:
> With regards to actually profiling, the amount of calls to malloc/free inside the clipping functions has gone from ~40% of execution time to 16%.

Move the Vector out of the function and into the class and clear them instead, then the number will drop to 0 :^)